### PR TITLE
Support install rules in the CMake build

### PR DESCRIPTION
**Pull request by petrhosek**
This allows using ruy from other projects through find_package.

### CODE REVIEW

File: ruy/CMakeLists.txt [hunk 1]
```diff
@@ -416,7 +416,7 @@ elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL s390 OR CMAKE_SYSTEM_PROCESSOR STREQUAL s
 elseif(CMAKE_SYSTEM_NAME STREQUAL Fuchsia)
   set(ruy_6_cpuinfo "")
 else()
-  set(ruy_6_cpuinfo "cpuinfo")
+  set(ruy_6_cpuinfo "cpuinfo::cpuinfo")
```
Comment (bjacob) [inline ruy/CMakeLists.txt:419]: This file is generated - as explained in the comment at the top. If it needs to be updated, please to try to achieve that by updating and rerunning the generator script instead.

Please also help me understand why this change is needed - as you added those `add_library(cpuinfo::cpuinfo ALIAS cpuinfo)` in CMakeLists.txt, that sounded like that was taking care of that?

File: cmake/ruy_include_directories.cmake [hunk 2]
```diff
@@ -14,20 +14,8 @@
...
-        "${PROJECT_SOURCE_DIR}/third_party/cpuinfo/include"
-    )
-  endif()
-  if ((gtest IN_LIST DEPS) OR
-      (gtest_main IN_LIST DEPS))
-    target_include_directories(${NAME}
-      PRIVATE
-        "${PROJECT_SOURCE_DIR}/third_party/googletest/googletest"
-    )
-  endif()
```
Comment (bjacob) [inline cmake/ruy_include_directories.cmake:32]: Why is this not needed anymore?

File: cmake/ruyConfig.cmake.in [hunk 3]
```diff
@@ -0,0 +1,7 @@
+include(CMakeFindDependencyMacro)
```
Comment (bjacob) [inline cmake/ruyConfig.cmake.in:1]: Can you please add a comment at the start of this file explaining its purpose? How do you plan to consume this file?

File: CMakeLists.txt [hunk 4]
```diff
@@ -35,14 +35,32 @@ endif()
...
+include(CMakeDependentOption)
+
+option(RUY_USE_EXTERNAL_CPUINFO "Use external cpuinfo" OFF)
+
+cmake_dependent_option(RUY_FIND_CPUINFO "Use find_package to find cpuinfo" ON "RUY_USE_EXTERNAL_CPUINFO" OFF)
+
 # Skip cpuinfo if it was already generated, which can happen when ruy is
 # a subdirectory in a wider project that already uses cpuinfo.
-if (NOT TARGET cpuinfo)
+if (RUY_USE_EXTERNAL_CPUINFO)
```
Comment (bjacob) [inline CMakeLists.txt:53]: The existing comment, left unchanged in this diff, referred specifically to the existing `if (NOT TARGET cpuinfo)` and is now out of place with the new `if (RUY_USE_EXTERNAL_CPUINFO)`.

Question: if `TARGET cpuinfo` is already `TRUE`, do you still want to go into this `if (RUY_USE_EXTERNAL_CPUINFO)`? If not, maybe you could reinstate the `if (NOT TARGET cpuinfo)`  and move the `if (RUY_USE_EXTERNAL_CPUINFO)` inside it? That would make it easier to review this code to ensure that it is a no-op for existing users.

File: CMakeLists.txt [hunk 5]
```diff
@@ -35,14 +35,32 @@ endif()
...
+  if (RUY_FIND_CPUINFO)
+    find_package(cpuinfo REQUIRED)
+  elseif (NOT TARGET cpuinfo::cpuinfo)
+    add_library(cpuinfo::cpuinfo ALIAS cpuinfo)
```
Comment (bjacob) [inline CMakeLists.txt:57]: That this `add_library` is made in 2 separate lines suggests at least asking: could this logic be simplified in some way so that this `add_library` would be needed only once?

File: ruy/CMakeLists.txt (hunk 1 above)
Comment (petrhosek) [reply ruy/CMakeLists.txt:419]: `add_library(cpuinfo::cpuinfo ALIAS cpuinfo)` works the other way round, it defines `cpuinfo::cpuinfo` as an alias for `cpuinfo` in the case where cpuinfo is used as subdirectory; when used through `find_package`, `cpuinfo-targets.cmake` defines the `cpuinfo::cpuinfo` target.

This is the idiomatic way to consume CMake packages, see for example https://github.com/abseil/abseil-cpp/blob/1963f10ae5cb32a7ea6d96b928f69d3c7fba0139/CMakeLists.txt#L140

File: cmake/ruy_include_directories.cmake (hunk 2 above)
Comment (petrhosek) [reply cmake/ruy_include_directories.cmake:32]: I'm not sure why this was needed in the first place since both `cpuinfo` and `gtest` targets set the include directory property which applies to their dependents so this should be unnecessary.

File: cmake/ruyConfig.cmake.in (hunk 3 above)
Comment (petrhosek) [reply cmake/ruyConfig.cmake.in:1]: I have added the comment, this follows the CMake conventions for packages and will be picked up automatically by `find_package(ruy)`. See https://github.com/abseil/abseil-cpp/blob/1963f10ae5cb32a7ea6d96b928f69d3c7fba0139/CMake/abslConfig.cmake.in for an example of an equivalent file in Abseil.

File: CMakeLists.txt (hunk 4 above)
Comment (petrhosek) [reply CMakeLists.txt:53]: I aimed to provide a more explicit control over the behavior but you're correct that this changes the behavior for existing users which may be undesirable. I have reinstated (and updated) the original condition.

File: cmake/ruyConfig.cmake.in (hunk 3 above)
Comment (bjacob) [reply cmake/ruyConfig.cmake.in:3]: Thanks!

File: CMakeLists.txt (hunk 4 above)
Comment (bjacob) [reply CMakeLists.txt]: Thanks, I can more easily make sense for it now. And I've tried it out, it works for me.
